From ab14b70fda8e1932702334762a60f37ba20ec65d Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Mon, 12 Sep 2016 21:11:22 +0100
Subject: [PATCH 5/9] daemon/endsession: Correctly handle changes to
 compositing state

Per issue #581, the End Session Dialog no longer has shadows, or indeed,
transparency, due to setting a null RGBA visual during start up. This is
because we're starting up before the WM is fully initialised and we don't
have an RGBA visual, due to a lack of composite extension.

We now correctly track the changes and install the RGBA visual the first
chance we get.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 daemon/endsession.vala | 34 ++++++++++++++++++++++++++++++----
 1 file changed, 30 insertions(+), 4 deletions(-)

diff --git a/daemon/endsession.vala b/daemon/endsession.vala
index 974b90d..2d8a66c 100644
--- a/daemon/endsession.vala
+++ b/daemon/endsession.vala
@@ -98,6 +98,34 @@ public class EndSessionDialog : Gtk.Window
         }
     }
 
+    /**
+     * Attempt to set the RGBA visual
+     */
+    [DBus (visible = false)]
+    private void on_realized()
+    {
+        Gdk.Visual? visual = screen.get_rgba_visual();
+        if (visual != null) {
+            this.set_visual(visual);
+        }
+    }
+
+    /**
+     * Update the RGBA visual if its available when compositing changes
+     * This is required as we may be constructed before the window manager
+     * springs into life
+     */
+    [DBus (visible = false)]
+    private void on_composite_changed()
+    {
+        Gdk.Visual? visual = screen.get_rgba_visual();
+        if (visual != null) {
+            this.set_visual(visual);
+        } else {
+            this.set_visual(screen.get_system_visual());
+        }
+    }
+
     [DBus (visible = false)]
     public EndSessionDialog()
     {
@@ -107,10 +135,8 @@ public class EndSessionDialog : Gtk.Window
         set_has_resize_grip(false);
         set_resizable(false);
 
-        Gdk.Visual? visual = screen.get_rgba_visual();
-        if (visual != null) {
-            this.set_visual(visual);
-        }
+        realize.connect(on_realized);
+        screen.composited_changed.connect(on_composite_changed);
 
         var header = new Gtk.EventBox();
         set_titlebar(header);
-- 
2.10.0

